CL-6867 Fix password reset emails sometimes being slow to send out

nTraum opened the pull request:
## Checklist

- [ ] Added entry to changelog
<details>
<summary>More info</summary>
Add a concise line to the 'Next release' section of the changelog (docs/README.md) so people other than developers can understand what has changed where. E.g. 'Added an error message to the project name field of the project edit form (Admin > Projects > Edit)'.
</details>

- [ ] WCAG 2.1 AA proof
<details>
<summary>More info</summary>
For front-end devs only. Is your work conforming with the WCAG 2.1 AA rules? If you need more info, read the [a11y page](https://www.notion.so/citizenlab/a11y-7568f83d42ab4895ac133b89d358997b) on our Notion.
</details>

- [ ] Tests
<details>
<summary>More info</summary>

### Unit tests

Did you add relevant unit tests?

### E2E tests

Sometimes it can be more efficient to update E2E tests after CI has run them. If you know which ones to update, go ahead! E2E template cl2-back:

```bash
docker compose run --rm web bin/rails cl2_back:create_tenant[localhost,e2etests_template]
```

</details>

- [ ] Prepared branch for code review
<details>
<summary>More info</summary>
Reviewed code to reduce unnecessary back and forth (removal of console.log, comments, ...)? Added comments to clarify code, emphasize what to pay attention to, etc.?
</details>

## Links

- [citizenlab-ee PR](**put URL here** or remove)
- [Specs](**put URL here** or remove)
- [Epic Deployment](**put URL here** or remove)

## How urgent is a code review?

Let the reviewer(s) know how urgent the code review is, so they can prioritize their work accordingly. Be specific (e.g. by Wednesday, end of the day/this week/... is better than 'urgent' or 'very urgent'). Optionally provide a word of explanation on your deadline.

nTraum committed:
Format ResetPasswordController and ResetPasswordService with Rubocop

nTraum committed:
Rename log_password_reset_to_segment to log_password_reset_activity

Which brings more clarity what the method does.

nTraum committed:
Add docs for ResetPasswordService

nTraum committed:
Make ResetPasswordService#secret private

It's not used outside the class.

nTraum committed:
Fix typo

nTraum commented:
At first I was suspecting Mailgun to be slow at delivering emails (they reported incidents on Nov 9th and 12th). I then looked at the implementation and noticed that the PasswordResetEmail is, as any other email on the backend, labeled and handled as a campaign email.

This in itself is not a problem yet, but it seems Que, our background processing engine, is sometimes struggling to keep up with all the new jobs:

the PasswordResetEmail is one of these background jobs and is not prioritized higher than any other (less important campaign email), which means when the queue is long, you have to wait for a very long time. It also seems we are batching lots of jobs on an hourly basis, I am not sure what the source of them is.


![image](https://user-images.githubusercontent.com/866318/142213561-a8d2987a-996b-4fe0-b768-2e2c7c15da69.png)

I then looked at how to prioritize the background job that is responsible for sending out the email, but the level of indirection makes this really hard. We are scheduling three background jobs that need to execute before the mail is actually sent!

```shell
WebApi::V1::ResetPasswordController#reset_password_email
-> ResetPasswordService.new.log_password_reset_activity
->-> LogActivityJob.set(wait: 2.seconds).perform_later(user, 'requested_password_reset', ...) # priority needed!
->->-> EmailCampaigns::TriggerOnActivityJob.perform_later(activity)  # priority needed!
->->->-> DeliveryService.new.send_on_activity(activity)
->->->->-> campaign.mailer_class.with(campaign: campaign, command: command) .campaign_mail .deliver_later(wait: command[:delay] || 0) # priority needed!
```

So we would need to prioritize three different background jobs higher than default, but two of them are generic jobs that are not specific to password resets, but handle generic cases: `LogActivityJob`, `TriggerOnActivityJob`.

I wonder what the right approach here is. I think in the long run we definitely need to start prioritizing our que background jobs, because there will be times when que won't keep up with the number of jobs in the queue, but the current level of indirection makes this difficult.

I also don't fully understand why we need to go through the whole EmailCampaign logic for a simple PasswordReset email, which isn't even a campaign email, but a transactional email that is linked to one specific user and action. Should we simplify the code and remove it or is this actually needed?

@sebastienhoorens can you share your thoughts on this, I think you have worked on it before.

sebastienhoorens commented:
Yes, I completely agree. We're having issues with the Que sometimes and can't monitor it well at the moment, and as you say, there are different jobs involved which makes it hard to have the email send out faster, even if we increase the priority for some jobs. I still think the ideal solution would be to thoroughly improve how jobs are queued and prioritized, but this is probably too much effort.

For the user verification email, I see that the email is sent out right away, so, as you suggest, you could consider sending out the email right away without enqueueing any job. I think an improvement on this solution would be to enqueue one job with high priority. The advantage would be that this doesn't keep the API request pending to respond, the disadvantage is that you can't confirm to the user that the password reset email has successfully been sent out (which I think is why user verification works that way).

Perhaps we could introduce the concept of high priority (essential) emails or something like that, as opposed to email campaigns? That way we could apply this for all the important emails, such as password reset, user verification and the invitation email. I think it would be nice to have all those emails and how they are sent being implemented in the same manner.

nTraum commented:
Good thoughts, thanks for sharing them. :)

A quick note: I don't favour sending out the email right away in a synchronous way. We can have a background job that does that, but I would like the `ResetPasswordService` to enqueue this mailer job directly, instead of:

- First Enqueue a LogActivityJob
- Then Enqueue a TriggerOnActivityJob
- Then Enqueue a MailerClass

because in order to prioritize the email higher, we'd need to add prioritization to the preceding jobs as well. As these are generic jobs that are used in other scenarios as well, we would have to add complexity on when they are high priority and when they aren't, further complicating our setup. I am not if this is worth it and if there is a better way.

I understand that campaign emails have different needs and need more logic to them, but simple transactional emails like a password reset currently suffer a bit from the complexity that is involved in my opinion.

nTraum committed:
Deprecate LogActivityJob for LogActivityService

They both feature the same implementation.
We need to keep the LogActivityJob around for some time because there might be jobs in the que queue that need to processed. Removing the class here already would make them fail hard, so let's do it in the next release instead.

nTraum committed:
Replace LogActivityJob with LogActivityService for invocations with no delay

nTraum committed:
Replace all occurrences of LogActivityJob with delay with LogActivityService that runs instantly.

nTraum commented:
Closing this PR and reopening a new one because of a long standing-bug where force-pushing a branch does not trigger all CIs anymore. :(
